source-bigquery-batch/driver.go

```diff
@@ -0,0 +1,698 @@
+	log.WithFields(log.Fields{
+		"name":   res.Name,
+		"tmpl":   res.Template,
+		"cursor": res.Cursor,
+		"poll":   res.PollInterval,
+	}).Info("starting worker")
+
+	var queryTemplate, err = template.New("query").Parse(res.Template)
+	if err != nil {
+		return fmt.Errorf("error parsing template: %w", err)
+	}
+
+	for ctx.Err() == nil {
+		if err := c.poll(ctx, bindingIndex, queryTemplate, res); err != nil {
+			return fmt.Errorf("error polling table: %w", err)
+		}
+	}
+	return ctx.Err()
+}
+
+func quoteIdentifier(name string) string {
+	return "`" + strings.ReplaceAll(name, "`", "\\`") + "`"
+}
+
+func quoteColumnName(name string) string { return quoteIdentifier(name) }
+
+func (c *capture) poll(ctx context.Context, bindingIndex int, tmpl *template.Template, res *Resource) error {
+	var state, ok = c.State.Streams[res.Name]
+	if !ok {
+		return fmt.Errorf("internal error: no state for stream %q", res.Name)
+	}
+	var cursorNames = state.CursorNames
+	var cursorValues = state.CursorValues
+
+	var quotedCursorNames []string
+	for _, cursorName := range cursorNames {
+		quotedCursorNames = append(quotedCursorNames, quoteColumnName(cursorName))
+	}
+	var templateArg = map[string]any{
+		"IsFirstQuery": len(cursorValues) == 0,
+		"CursorFields": quotedCursorNames,
+	}
+
+	// Polling interval can be configured per binding. If unset, falls back to the
+	// connector global polling interval.
+	var pollStr = c.Config.Advanced.PollInterval
+	if res.PollInterval != "" {
+		pollStr = res.PollInterval
+	}
+	pollInterval, err := time.ParseDuration(pollStr)
+	if err != nil {
+		return fmt.Errorf("invalid poll interval %q: %w", res.PollInterval, err)
+	}
+
+	// Sleep until it's been more than <pollInterval> since the last iteration,
+	// then update the "Last Polled" timestamp.
+	if !state.LastPolled.IsZero() && time.Since(state.LastPolled) < pollInterval {
+		var sleepDuration = time.Until(state.LastPolled.Add(pollInterval))
+		log.WithFields(log.Fields{
+			"name": res.Name,
+			"wait": sleepDuration,
+			"poll": pollInterval,
+		}).Info("waiting for next poll")
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(sleepDuration):
+		}
+	}
+	log.WithField("name", res.Name).Info("ready to poll")
+	var pollTime = time.Now().UTC()
+	state.LastPolled = pollTime
+
+	// Acquire mutex so that only one worker at a time can be actively executing a query.
+	c.Mutex.Lock()
+	defer c.Mutex.Unlock()
+
+	var queryBuf = new(strings.Builder)
+	if err := tmpl.Execute(queryBuf, templateArg); err != nil {
+		return fmt.Errorf("error generating query: %w", err)
+	}
+	var query = queryBuf.String()
+
+	log.WithFields(log.Fields{"query": query, "args": cursorValues}).Info("executing query")
+	var q = c.DB.Query(query)
+	var params []bigquery.QueryParameter
+	for idx, val := range cursorValues {
+		params = append(params, bigquery.QueryParameter{
+			Name:  fmt.Sprintf("p%d", idx),
+			Value: val,
+		})
+	}
+	q.Parameters = params
+
+	rows, err := q.Read(ctx)
```

Comment:
Not requesting anything different here right now, but I'm betting this is going to be relatively slow, for reasons similar to those [noted for reading loads](https://github.com/estuary/connectors/issues/968) of `materialize-bigquery`. Fortunately, I have seen that the slowness when running in our production environment isn't nearly as bad as when I've tested on a local stack, but it's still not great. Enabling the [storage read API](https://pkg.go.dev/cloud.google.com/go/bigquery#Client.EnableStorageReadClient) makes reading results _much_ faster when I've tested it, but is pretty brutal on memory use for the connector. It could be something we need to look into more down the line.

Reply:
> Enabling the [storage read API](https://pkg.go.dev/cloud.google.com/go/bigquery#Client.EnableStorageReadClient) makes reading results much faster when I've tested it, but is pretty brutal on memory use for the connector

My experience has generally been that trading off increased memory usage for throughput is a bad move -- nearly all of the work getting SQL CDC memory usage under control boiled down to "remove some buffering that I previously added as a throughput optimization", so I'd say we shouldn't do that unless/until we have evidence that it's not fast enough as-is for real-world uses.